### PR TITLE
Stop using cached component in _analyze API

### LINKED ISSUE #19827: Stop using cached components in the _analyze API
We'd like to simplify `AnalysisService` and remove most of its members (#19814), maybe reducing it to just a `Map<String, Analyzer>`. To do that, we should remove calls to `tokenizer`, `charFilter`, and `tokenFilter` from the `_analyze` API, instead rebuilding these analyzers on the fly. This will make some calls to the `_analyze` API slower but it'll reduce the per index heap overhead.

### PR DESCRIPTION
**Pull request by johtani**
Stop calling tokenizer/tokenFilters/charFilter method of IndexService
Add some getAnalysisProvider methods
Change SynonymTokenFilterFactory constructor

Closes #19827

### CODE REVIEW

**Comment by nik9000** [comment]
I'll review it!

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 1]
```diff
@@ -136,14 +147,54 @@ public AnalysisService build(IndexSettings indexSettings) throws IOException {
...
-        tokenFilters.put("synonym", requriesAnalysisSettings((is, env, name, settings) -> new SynonymTokenFilterFactory(is, env, tokenizerFactories, name, settings)));
+        tokenFilters.put("synonym", requriesAnalysisSettings((is, env, name, settings) -> new SynonymTokenFilterFactory(is, env, this, name, settings)));
         final Map<String, TokenFilterFactory> tokenFilterFactories = buildMapping(false, "tokenfilter", indexSettings, tokenFiltersSettings, Collections.unmodifiableMap(tokenFilters), prebuiltAnalysis.tokenFilterFactories);
         final Map<String, AnalyzerProvider<?>> analyzierFactories = buildMapping(true, "analyzer", indexSettings, analyzersSettings,
                 analyzers, prebuiltAnalysis.analyzerProviderFactories);
         return new AnalysisService(indexSettings, analyzierFactories, tokenizerFactories, charFilterFactories, tokenFilterFactories);
     }
 
 
+    public AnalysisProvider<TokenizerFactory> getTokenizerProvider(String tokenizer, IndexSettings indexSettings) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: I think it'd be nice to have javadoc for these so it is clear to the casual observer that they are for building the provider in the context of some index's settings.

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 2]
```diff
@@ -70,6 +73,14 @@ public AnalysisRegistry(Environment environment,
         this.analyzers = unmodifiableMap(analyzers);
     }
 
+    public static Settings getSettingsFromIndexSettings(IndexSettings indexSettings, String groupName) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: Javadoc would be nice here too!

**Comment by nik9000** [comment]
LGTM - I asked for some javadoc but I'm happy for you to add that javadoc and merge whenever you get the chance!

**Comment by johtani** [comment]
@nik9000 Thx for reviewing! I added the javadoc.

**Comment by nik9000** [comment]
Thanks for the cleanup!

**Pull request merged**